### ISSUE TITLE
scx_layered failing to build with 6.7.0rc4

### ISSUE THREAD

**Issue by ptr1337** [sched-ext/sched_ext#90]
Hi,

I have tested several revs of this patchset (lkml, master) and all of them are throwing the same error.
I think there was probably a breaking change.

```
   Compiling scx_layered v0.0.1 (/home/ptr1337/projects/kernel/linux-cachyos/linux-cachyos-rc/src/linux-6.7-rc4/tools/sched_ext/scx_layered)
error: failed to run custom build command for `scx_layered v0.0.1 (/home/ptr1337/projects/kernel/linux-cachyos/linux-cachyos-rc/src/linux-6.7-rc4/tools/sched_ext/scx_layered)`

Caused by:
  process didn't exit successfully: `/home/ptr1337/projects/kernel/linux-cachyos/linux-cachyos-rc/src/linux-6.7-rc4/tools/sched_ext/build/release/build/scx_layered-a5071e99410c4263/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-changed=src/bpf/layered.h
  cargo:rerun-if-changed=/usr/lib/clang/16/include/stdbool.h
  cargo:rerun-if-changed=src/bpf/../../../ravg.bpf.h

  --- stderr
  clang
  thread 'main' panicked at build.rs:57:10:
  called `Result::unwrap()` on an `Err` value: Build(Failed to compile ./src/bpf/.output/layered.bpf.o from ./src/bpf/layered.bpf.c

  Caused by:
      0: Command `clang -g -D__TARGET_ARCH_x86 -mlittle-endian -I/home/ptr1337/projects/kernel/linux-cachyos/linux-cachyos-rc/src/linux-6.7-rc4/tools/sched_ext/build/include -I/home/ptr1337/projects/kernel/linux-cachyos/linux-cachyos-rc/src/linux-6.7-rc4/tools/sched_ext -I/home/ptr1337/projects/kernel/linux-cachyos/linux-cachyos-rc/src/linux-6.7-rc4/tools/include/uapi -I../../include -idirafter /usr/lib/clang/16/include -idirafter /usr/local/include -idirafter /usr/include -Wall -Wno-compare-distinct-pointer-types -O2 -mcpu=v3 -I/tmp/.tmpa6R3xI/bpf/src -fno-stack-protector -g -O2 -target bpf -c ./src/bpf/layered.bpf.c -o ./src/bpf/.output/layered.bpf.o` failed (exit status: 1)
      1: ./src/bpf/layered.bpf.c:277:26: error: no member named 'thread_group' in 'struct task_struct'; did you mean 'thread_node'?
                         p = container_of(next->thread_group.next, struct task_struct, thread_group);
                                                ^~~~~~~~~~~~
                                                thread_node
         /home/ptr1337/projects/kernel/linux-cachyos/linux-cachyos-rc/src/linux-6.7-rc4/tools/sched_ext/build/include/bpf/bpf_helpers.h:92:27: note: expanded from macro 'container_of'
                         void *__mptr = (void *)(ptr);                   \
                                                 ^
         /home/ptr1337/projects/kernel/linux-cachyos/linux-cachyos-rc/src/linux-6.7-rc4/tools/sched_ext/build/include/vmlinux.h:1481:19: note: 'thread_node' declared here
                 struct list_head thread_node;
                                  ^
         ./src/bpf/layered.bpf.c:277:65: error: no member named 'thread_group' in 'struct task_struct'; did you mean 'thread_node'?
                         p = container_of(next->thread_group.next, struct task_struct, thread_group);
                                                                                       ^~~~~~~~~~~~
                                                                                       thread_node
         /home/ptr1337/projects/kernel/linux-cachyos/linux-cachyos-rc/src/linux-6.7-rc4/tools/sched_ext/build/include/bpf/bpf_helpers.h:93:37: note: expanded from macro 'container_of'
                         ((type *)(__mptr - offsetof(type, member)));    \
                                                           ^
         /home/ptr1337/projects/kernel/linux-cachyos/linux-cachyos-rc/src/linux-6.7-rc4/tools/sched_ext/build/include/bpf/bpf_helpers.h:86:62: note: expanded from macro 'offsetof'
         #define offsetof(type, member)  ((unsigned long)&((type *)0)->member)
                                                                       ^
         /home/ptr1337/projects/kernel/linux-cachyos/linux-cachyos-rc/src/linux-6.7-rc4/tools/sched_ext/build/include/vmlinux.h:1481:19: note: 'thread_node' declared here
                 struct list_head thread_node;
                                  ^
         2 errors generated.
         )
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
make: *** [Makefile:218: scx_layered] Error 101

```

**Comment by htejun** [comment]
Hahaha, it's saying task_struct.thread_group is not there. That's not an optional field in the upstream kernels at least. Can you attach the result of `pahole vmlinux`? Let's see what's going on with the task_struct definition. Also, where would I find the source for your kernel?

**Comment by ptr1337** [comment]
You can find the source here, its a PKGBUILD:
https://github.com/CachyOS/linux-cachyos/blob/master/linux-cachyos-sched-ext/PKGBUILD#L24

We run the build process of the kernel, its headers and sched-ext scheduler in one PKGBUILD, which makes it quite good to maintain updates related to sched-ext.
This is the used patchset:
https://github.com/CachyOS/kernel-patches/blob/master/6.7/sched/0001-sched-ext.patch

This is basically the the latest lkml sched-ext + nest_scheduler.

Changes between rc3 and rc4 sched-ext patchset is just following:
https://github.com/CachyOS/kernel-patches/commit/dc3719ac87ebd6e1c39da1b7a90991f8d3a153d5#diff-398b052620237d48be7b303e07f6fd37020e19704a674f29a013d3af13d4a4fd

It does only include the latest nest scheduler changes. rc3 sched-ext worked fine.

I have attached you the "pahole.log".
[pahole.log](https://github.com/sched-ext/sched_ext/files/13550163/pahole.log)

**Comment by htejun** [comment]
I see. 8e1f385104ac ("kill task_struct->thread_group") got rid of it. The same iteration can be done by walking task->signal->thread_head. Will update the code. Thanks.

**Comment by htejun** [comment]
@ptr1337 We just created https://github.com/sched-ext/scx repo which will be the source of truth for scx schedulers. tools/sched_ext will be synced back from there but we might whittle it down a bit in the future.

https://github.com/sched-ext/scx/pull/2 should fix the reported problem. Once it gets merged there, I'll sync it back to this repo. Thanks.

**Comment by ptr1337** [comment]
Thanks @htejun @Decave 
I will test it tomorrow with the patch and report back

**Comment by htejun** [comment]
Synced the fix through #91. I'll close it for now. Please feel free to reopen if it's not fixed. Thanks.
